newIDE/app/src/VariablesList/VariablesList.js

```diff
@@ -0,0 +1,1360 @@
+    setExpandedNodes([
+      ...inheritedExpandedNodes,
+      ...getExpandedNodeIdsFromVariablesContainer(props.variablesContainer),
+    ]);
+  };
+
+  const DragSourceAndDropTarget = React.useMemo(
+    () => makeDragSourceAndDropTarget('variable-editor'),
+    []
+  );
+
+  const canDrop = (nodeId: string): boolean => {
+    if (nodeId.startsWith(inheritedPrefix)) return false;
+    const { current } = draggedNodeId;
+    if (!current) return false;
+
+    const targetVariableContext = getVariableContextFromNodeId(
+      nodeId,
+      props.variablesContainer
+    );
+    const { lineage: targetLineage } = targetVariableContext;
+
+    const draggedVariableContext = getVariableContextFromNodeId(
+      current,
+      props.variablesContainer
+    );
+    const { variable: draggedVariable } = draggedVariableContext;
+
+    const targetLineageVariables = targetLineage.map(
+      context => context.variable
+    );
+    if (targetLineageVariables.includes(draggedVariable)) return false;
+
+    const movementType = getMovementTypeWithinVariablesContainer(
+      draggedVariableContext,
+      targetVariableContext
+    );
+
+    switch (movementType) {
+      case 'InsideTopLevel':
+      case 'TopLevelToStructure':
+      case 'StructureToTopLevel':
+      case 'FromStructureToAnotherStructure':
+      case 'FromArrayToAnotherArray':
+      case 'InsideSameArray':
+        return true;
+      case 'FromStructureToArray':
+      case 'FromArrayToStructure':
+      case 'ArrayToTopLevel':
+      case 'InsideSameStructure':
+      default:
+        return false;
+    }
+  };
+
+  const dropNode = (nodeId: string): void => {
+    if (nodeId.startsWith(inheritedPrefix)) return;
+    const { current } = draggedNodeId;
+    if (!current) return;
+
+    // TODO: Add logic to copy dragged variable instead of moving it if Alt/Opt key is pressed
+    // React-dnd keeps the focus when user is dragging so keyboard shortcut instance
+    // cannot detect if the key is pressed while dragging. React-dnd has issues to
+    // return event data about pressed keys when mouse is up.
+
+    const targetVariableContext = getVariableContextFromNodeId(
+      nodeId,
+      props.variablesContainer
+    );
+    const { lineage: targetLineage, name: targetName } = targetVariableContext;
+    const targetVariableParentVariable = getDirectParentVariable(targetLineage);
+    if (!targetName) return;
+
+    const draggedVariableContext = getVariableContextFromNodeId(
+      current,
+      props.variablesContainer
+    );
+    const {
+      variable: draggedVariable,
+      lineage: draggedLineage,
+      name: draggedName,
+    } = draggedVariableContext;
+    const draggedVariableParentVariable = getDirectParentVariable(
+      draggedLineage
+    );
+    if (!draggedVariable || !draggedName) return;
+
+    const targetLineageVariables = targetLineage.map(
+      context => context.variable
+    );
+    if (targetLineageVariables.includes(draggedVariable)) return;
```

Comment:
This could almost be extracted to a well named function

Reply:
(to show in one line that you care about checking if the variable is dragged inside its descendance, which won't work)